From 30d1e8e1893b42eec8ca0c38b871214e4055e8dc Mon Sep 17 00:00:00 2001
From: Anmol Jain <ajain@ciq.com>
Date: Fri, 8 Aug 2025 17:22:24 +0000
Subject: [PATCH] move-size_t-decl-outside-for-loop

---
 modules/pam_namespace/pam_namespace.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/modules/pam_namespace/pam_namespace.c b/modules/pam_namespace/pam_namespace.c
index 76fc33a..3ced97e 100644
--- a/modules/pam_namespace/pam_namespace.c
+++ b/modules/pam_namespace/pam_namespace.c
@@ -868,7 +868,8 @@ static int parse_config_file(struct instance_data *idata)
     }
 
     if (filename_list != NULL) {
-	for (size_t i = 0; filename_list[i] != NULL; i++)
+	size_t i;
+	for (i = 0; filename_list[i] != NULL; i++)
 	    free(filename_list[i]);
 	free(filename_list);
     }
-- 
2.43.5

